### PR TITLE
return the original query planner error and add extensions code

### LINKED ISSUE #1560: potential improvements to error shape when validation error occurs
Currently our `errors` responses when there is a query validation error includes the various layers of error-prefixing.  For example, the error returned to the client when an unknown directive is included on a query:
 
```json5
{"errors":[{"message":"value retrieval failed: couldn't plan query: query validation errors: Directive \"@myDirective\" may not be used on FIELD.","locations":[],"path":null}]}
```

While it could be worth having that full `value retrieval failed: couldn't plan query: query validation errors: ` in our debugging logs, I think it makes less sense as a publicly returned error and more like a leaky abstraction of our internal error messages (which could change over time!).

In addition to being more concise, I think removing the prefix would create a more user-friendly message for the client-consumers.  For example, we might just shorten this to:

```json5
{"errors":[{"message":"Directive \"@myDirective\" may not be used on FIELD.","locations":[],"path":null}]}
```

Taking this a step further though, we could also make the error messages more consistent with what Apollo Server currently returns (though I'm curious what validation errors from other servers look like!), including a valuable machine-readable error message property called `code`.  For example, here's what a similar error message looks like on Apollo Server:

```json5
{
  "errors": [
    {
      "message": "Unknown directive \"@myDirective\".",
      "extensions": {
        "code": "GRAPHQL_VALIDATION_FAILED",
        "exception": {
          "stacktrace": [
            "GraphQLError: Unknown directive \"@myDirective\".",
            "    at Object.Directive (/home/projects/github-zz8rcx/node_modules/graphql/validation/rules/KnownDirectivesRule.js:56:11)",
            "    at Object.enter (/home/projects/github-zz8rcx/node_modules/graphql/language/visitor.js:301:32)",
            "    at Object.enter (/home/projects/github-zz8rcx/node_modules/graphql/utilities/TypeInfo.js:391:27)",
            "    at visit (/home/projects/github-zz8rcx/node_modules/graphql/language/visitor.js:197:21)",
            "    at validate (/home/projects/github-zz8rcx/node_modules/graphql/validation/validate.js:91:24)",
            "    at validate (/home/projects/github-zz8rcx/node_modules/apollo-server-core/dist/requestPipeline.js:186:39)",
            "    at processGraphQLRequest (/home/projects/github-zz8rcx/node_modules/apollo-server-core/dist/requestPipeline.js:98:34)",
            "    at async processHTTPRequest (/home/projects/github-zz8rcx/node_modules/apollo-server-core/dist/runHttpQuery.js:220:30)"
          ]
        }
      }
    }
  ]
}
```

Having the `code` of `GRAPHQL_VALIDATION_FAILED` also makes it easier for clients and tooling to programmatically figure out what is wrong with their request.   The full list of those codes we offer in Apollo Server [can be found here](https://www.apollographql.com/docs/apollo-server/data/errors/).

Of course, we don't need to include a JavaScript stacktrace.  (In fact, those aren't returned in Production anyhow, but I didn't run this in production.). I think it would be sufficient to leave off the whole `exception` object in Rust-land.

Thoughts?

### PR DESCRIPTION
**Pull request by bnjjj**
TLDR I'm not very proud of what I did here. It's very hard to return specific errors or just catch an error in our codebase due to our usage of `BoxError`. That's why I had to downcast and so one ...

I think in a long term we should think about a better way to handle errors. Let's discuss about it.

close #1560

### CODE REVIEW

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/Cargo.toml [hunk 1]
```diff
@@ -120,7 +120,8 @@ reqwest = { version = "0.11.11", default-features = false, features = [
     "json",
     "stream",
 ] }
-router-bridge = "0.1.3"
+# router-bridge = "0.1.3"
+router-bridge = { git = "https://github.com/apollographql/federation-rs.git", branch = "garypen/1519-slghtly-better-errors"}
```
Comment (garypen) [inline apollo-router/Cargo.toml:124]: I presume this fix is going to wait for 0.1.4?

File: apollo-router/src/error.rs [hunk 2]
```diff
@@ -136,10 +137,19 @@ impl FetchError {
 #[serde(rename_all = "camelCase")]
 pub struct Location {
     /// The line number.
-    pub line: i32,
+    pub line: u32,
```
Comment (garypen) [inline apollo-router/src/error.rs:140]: Nice catch!

File: apollo-router/src/error.rs [hunk 3]
```diff
@@ -136,10 +137,19 @@ impl FetchError {
 #[serde(rename_all = "camelCase")]
 pub struct Location {
```
Comment (garypen) [inline apollo-router/src/error.rs:138]: Why don't we just use the `router_bridge` Location here? Is it because of the serde rename requirement? We could probably put that into the router bridge and it would be fine.

If we did just use the same type, then it would help avoid errors such as you caught just below.

File: apollo-router/tests/snapshots/integration_tests__defer_path_with_disabled_config.snap [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
 source: apollo-router/tests/integration_tests.rs
+assertion_line: 558
 expression: only
 ---
 {
   "errors": [
     {
-      "message": "value retrieval failed: couldn't plan query: query validation errors: Unknown directive \"@defer\".",
-      "path": null
+      "message": "Unknown directive \"@defer\".",
```
Comment (garypen) [inline apollo-router/tests/snapshots/integration_tests__defer_path_with_disabled_config.snap:9]: Nice!

**Review by garypen — APPROVED** [review]
Looks great! I asked a question about `Location`, but we could ignore here and file a follow-up issue I think.

**Comment by garypen** [comment]
fixes: #1511

File: apollo-router/Cargo.toml (hunk 1 above)
Comment (bnjjj) [reply apollo-router/Cargo.toml:124]: yes

**Comment by bnjjj** [comment]
I deleted the `Location` type to only use the `Location` type from the router-bridge. Thanks for your reviews.
